geoutils/profiler.py

```diff
@@ -0,0 +1,417 @@
+        def my_function():
+
+    """
+
+    def decorator_generator(func):  # type: ignore
+        """
+        Inner function
+        """
+
+        def wrapper_profile(*args, **kwargs):  # type: ignore
+            """
+            Profiling wrapper
+
+            Generate profiling logs of function, run
+
+            :return: func(*args, **kwargs)
+            """
+            # if profiling is disabled, remove overhead
+
+            if not Profiler.enabled:
+                return func(*args, **kwargs)
+
+            func_name = name
+            if Profiler.selection_activated and name not in Profiler.functions_selected:
+                print(Profiler.selection_activated, Profiler.functions_selected, name)
+                return func(*args, **kwargs)
+
+            uuid_function = str(uuid.uuid4())
+            uuid_parent = Profiler.running_processes[-1] if Profiler.running_processes else "__main__"
+            level = len(Profiler.running_processes)
+
+            if name is None:
+                func_name = func.__name__.capitalize()
+
+            Profiler.running_processes.append(uuid_function)
+
+            if memprof:
+                # Launch memory profiling thread
+                child_pipe, parent_pipe = Pipe()
+                thread_monitoring = MemProf(os.getpid(), child_pipe, interval=interval)
+                thread_monitoring.start()
+                if parent_pipe.poll(1):  # wait for thread to start
+                    parent_pipe.recv()
+
+            start_time = time.time()
+            res = func(*args, **kwargs)
+            total_time = time.time() - start_time
+
+            if memprof:
+                # end memprofiling monitoring
+                parent_pipe.send(0)
+
+            Profiler.running_processes.pop(-1)  # remove function from call list
+
+            func_data = {
+                "level": level,
+                "uuid_function": uuid_function,
+                "name": func_name,
+                "uuid_parent": uuid_parent,
+                "time": total_time,
+                "call_time": start_time,
+                "memory": thread_monitoring.mem_data if memprof else None,
+            }
+            Profiler.add_profiling_info(func_data)
+            return res
+
+        return wrapper_profile
+
+    return decorator_generator
+
+
+class MemProf(Thread):
+    """
+    MemProf
+
+    Profiling thread
+    """
+
+    def __init__(self, pid: int, pipe: connection.Connection, interval: float) -> None:
+        """
+        Init function of MemProf
+
+        :param pid: The process ID of the monitored process
+        :param pipe: The pipe used to send the end monitoring signal
+        :param interval: Time interval (seconds) between memory measurements
+        """
+        super().__init__()
+        self.pipe = pipe
+        self.interval = interval
+        self.process = psutil.Process(pid)
```

Comment:
Are you monitoring a single process? Will it work with multiprocessing or Dask enabled functions?